src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs

```diff
@@ -22,6 +22,10 @@ public override void Read(JsonTokenType tokenType, ref ReadStack state, ref Utf8
             {
                 // Forward the setter to the value-based JsonPropertyInfo.
                 JsonPropertyInfo propertyInfo = ElementClassInfo.GetPolicyProperty();
+                if (state.Current.IsProcessingDictionary && state.Current.KeyName == null)
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(RuntimePropertyType, reader, state.JsonPath);
```

Comment:
A question as much as a comment, in most of System.Text.Json there are returns immediately after calls to ThrowHelper. I assume this is because the JIT doesn't (can't?) know that the method never returns?
```suggestion
                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(RuntimePropertyType, reader, state.JsonPath);
                    return;
```

Reply:
Yes sometimes a `return` is necessary.

Throwhelper is there for perf - to avoid unnecessary branching in the IL in perf-sensitive areas.